x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtValidateUtil.java

```diff
@@ -0,0 +1,359 @@
+        final List<JWK> filterKid = JwkValidateUtil.removeJwkKidMismatches(jwkSet.getKeys(), jwt.getHeader().getKeyID());
+        final List<JWK> filterAlg = JwkValidateUtil.removeJwksAlgMismatches(filterKid, jwt.getHeader().getAlgorithm());
+        final List<JWK> filterUse = JwkValidateUtil.removeJwksKeyUseMismatches(filterAlg, KeyUse.SIGNATURE);
+        final List<JWK> filterOp = JwkValidateUtil.removeJwksKeyOperationMismatches(filterUse, KeyOperation.VERIFY);
+        final List<JWK> filterType = JwkValidateUtil.removeJwkTypeMismatches(filterOp, jwt.getHeader().getAlgorithm());
+        for (final JWK jwk : filterType) {
+            if (jwt.verify(JwtValidateUtil.createJwsVerifier(jwk))) {
+                return;
+            }
+        }
+        throw new Exception("Verify failed using " + filterAlg.size() + " of " + jwkSet.getKeys().size() + " provided JWKs.");
+    }
+
+    public static JWSVerifier createJwsVerifier(final JWK jwk) throws JOSEException {
+        if (jwk instanceof RSAKey rsaKey) {
+            return new RSASSAVerifier(rsaKey);
+        } else if (jwk instanceof ECKey ecKey) {
+            return new ECDSAVerifier(ecKey);
+        } else if (jwk instanceof OctetSequenceKey octetSequenceKey) {
+            return new MACVerifier(octetSequenceKey);
+        }
+        throw new JOSEException(
+            "Unsupported class ["
+                + (jwk == null ? "null" : jwk.getClass().getCanonicalName())
+                + "]. Supported classes are ["
+                + RSAKey.class.getCanonicalName()
+                + ", "
+                + ECKey.class.getCanonicalName()
+                + ", "
+                + OctetSequenceKey.class.getCanonicalName()
+                + "]."
+        );
+    }
+
+    public static JWSSigner createJwsSigner(final JWK jwk) throws JOSEException {
+        if (jwk instanceof RSAKey rsaKey) {
+            return new RSASSASigner(rsaKey);
+        } else if (jwk instanceof ECKey ecKey) {
+            return new ECDSASigner(ecKey);
+        } else if (jwk instanceof OctetSequenceKey octetSequenceKey) {
+            return new MACSigner(octetSequenceKey);
+        }
+        throw new JOSEException(
+            "Unsupported class ["
+                + (jwk == null ? "null" : jwk.getClass().getCanonicalName())
+                + "]. Supported classes are ["
+                + RSAKey.class.getCanonicalName()
+                + ", "
+                + ECKey.class.getCanonicalName()
+                + ", "
+                + OctetSequenceKey.class.getCanonicalName()
+                + "]."
+        );
+    }
+
+    public static boolean verifySignedJWT(final JWSVerifier jwtVerifier, final SignedJWT signedJwt) throws Exception {
+        return signedJwt.verify(jwtVerifier);
+    }
+
+    public static SignedJWT signSignedJwt(final JWSSigner jwtSigner, final JWSHeader jwsHeader, final JWTClaimsSet jwtClaimsSet)
```

Comment:
`signJwt`() instead ? If it's signed we shouldn't sign it again :)

Reply:
Renamed.